Max/comments integration

limmabean opened the pull request:
I need to fix the tag modal but I'm opening this up now because everything else is fixed.

karpawich committed:
feat: comments content service

karpawich committed:
feat: rough draft of Comments source

Commit:
Added Comment Card to Report Details Page

Commit:
SMTCTags contain Boolean field for enabling Comments

Commit:
Typo Fix in Schema File

karpawich committed:
Merge remote-tracking branch 'origin/ML-TagCommentAquistionField' into max/comments-integration

karpawich committed:
feat: link acquisition system with special tags

karpawich committed:
feat: only send API request for tagged CT reports

karpawich committed:
feat: associate comment reports with their post

karpawich committed:
feat: show content of original posts for comments

karpawich committed:
fix: fix  _media field typo

karpawich committed:
fix: properly complete URLs

karpawich committed:
fix: guarantee comments content service fetch cb

karpawich committed:
feat: add cross-env for running on Windows in dev mode

karpawich committed:
feat: persist lastFetched across runtimes

karpawich committed:
Revert "feat: add cross-env for running on Windows in dev mode"

This reverts commit 49938f477ce41cc2c29ab5188f72248d81ea11c6.

karpawich committed:
feat: fetch comments by their authoredAt date

karpawich committed:
fix: properly parse comment URL

karpawich committed:
fix: don't panic if comments status code is 409

karpawich committed:
chore: support Windows in dev mode

Commit:
Merge branch 'ML-Comments-ReportsPage' into ML/CommentQueryEndpoint

Commit:
Changes to the router and the API endpoint on the front-end

The API endpoint is Report.queryComments and it takes a page number and a id. The Show page lacks ability to select many comments and tag. But can flag and add incidents/remove incidents.

Commit:
Changes to the Back end report controller and fixed some frontend issues.

Also MAX you can overwrite report.js because I put a commentTo in there just to get it working for a test

Commit:
Removed Logging from frontend

Commit:
Added Id to the Row so that a link to the comment from the reports/batch page can link it directly to the comment on the screen.

karpawich committed:
Merge branch 'max/comments-integration' into ML/CommentQueryEndpoint

karpawich committed:
Merge pull request #392 from TID-Lab/ML/CommentQueryEndpoint

Ml/comment query endpoint

karpawich committed:
Merge pull request #393 from TID-Lab/develop

Latest changes

karpawich committed:
Merge branch 'max/comments-integration' of https://github.com/tid-lab/aggie into max/comments-integration

karpawich committed:
fix: variable typo in comments content service

karpawich committed:
fix: add missing return after callback

karpawich committed:
feat: move back to fetch by lastFetchDate

karpawich committed:
feat: match the comment acquisition API

karpawich committed:
fix: remove lastFetchedDate from sources

karpawich committed:
feat: persist lastFetchDate across runtimes

karpawich committed:
fix: pass in count to comments CS requests

karpawich committed:
fix: store lastReportDate at the beginning

karpawich committed:
fix: tell HTTP that the body is JSON

karpawich committed:
feat: use lastAuthoredDate for comments CS paging

karpawich committed:
feat: disable acquisition of group posts

karpawich committed:
tmp: debug comment paging

karpawich committed:
tmp: dehug comments paging #2

Commit:
Removed Authors from Comments, Added Thumbnails, Fixed Pagination

Commit:
Merge branch 'max/comments-integration' of https://github.com/TID-Lab/aggie into max/comments-integration

karpawich committed:
chore: delete debug lines for comment paging

Commit:
Merge branch 'develop' into max/comments-integration

# Conflicts:
#	lib/api/v1/report-controller.js

Commit:
Merge branch 'develop' into max/comments-integration

Commit:
Merge remote-tracking branch 'origin/max/comments-integration' into max/comments-integration

cooperka commented on lib/api/v1/report-controller.js:
```diff
@@ -151,10 +162,7 @@ module.exports = function(app, user) {
-            if (!res.headersSent) res.send(err.status, err.message)
-            return;
-          }
+          if (err) return res.send(err.status, err.message);
```
Are these all reverted because it's handled automatically now in https://github.com/TID-Lab/aggie/pull/398 or is that separate?

karpawich replied:
Separate. They were accidentally reverted, but I just reverted them back.

cooperka commented on config/secrets.json.example:
```diff
@@ -15,6 +15,12 @@
+    "username": "Enter username",
+    "password": "Enter password",
+    "baseUrl": "Enter base URL",
+    "count": 100
```
`count` is not very descriptive. It looks like it's only used a few places, could we be more verbose? (I'm not actually sure what it does after reading the code, since it's just sent as params to the API)

karpawich replied:
Yes. The ambiguous name is used in an effort to mimic the CrowdTangle content service so that the comments content service can be edited by Harshil, Daniel, or anyone else who helped to write the CT content service.

I've refactored the name to `pageCount` to reflect its role in telling the API the # of comments to return per page. However, I still want to keep the nomenclature similar so as to ensure others can make edits across both content services.

cooperka replied:
Thanks! As long as someone a year from now can look and figure out what it means. pageCount sounds reasonable.

cooperka commented on lib/fetching/content-services/comments-content-service.js:
```diff
@@ -0,0 +1,170 @@
+
+        const { acquiredAt, timestamp } = comments[comments.length - 1];
+
+        // TODO rewrite with async/await to avoid callback hell
```
😁 

(It's pretty straightforward to simplify this if you wanna take a swing at it, happy to send a quick example I have. Not blocking, just nice.)

karpawich replied:
Code rewritten to use async/await :)

cooperka commented on lib/fetching/content-services/comments-content-service.js:
```diff
@@ -0,0 +1,170 @@
+        metadata,
+    }
+
+    const post = this.cache[postUrl];
```
This memoization can be great for performance, but do you know if it's really necessary? Generally `findOne` should be very fast assuming reports are indexed on `url`. I worry this might be premature optimization that adds complexity without explicit benefit.

karpawich replied:
I am hesitant to remove the cache, for the reason that I think it will reduce what might be 100 `Report.findOne` calls to 1-3 on average. To explain further:

Assuming the comment acquisition system runs at full capacity, then `_parse` will be called 100 times every ten seconds for each of the 100 comments fetched from the API. On average, these 100 comments will be associated with just 1-3 posts. Thus, caching post reports fetched from the database will reduce 100 separate and mostly redundant indexed `Report.findOne` calls to 1-3.

@cooperka knowing the context better, is `Report.findOne` fast enough when indexed by URL such that even a search by URL through the hundreds of thousands of reports is trivial to execute? If so, I can happily remove this optimization.

cooperka replied:
Finding via index should generally be `O(1)` (extremely fast) in all cases. **Searching** (as in, arbitrary regex match) is slower but I don't think you're doing that here.

Anyway, I assume it's currently working so let's keep as is. Don't want to throw a wrench in things this close to the election!

karpawich replied:
I'll at least at a TODO note for the future. Did not know that indexed searches are O(1)!

cooperka commented on models/report.js:
```diff
@@ -85,11 +88,23 @@ schema.methods.addSMTCTag = function(smtcTagId) {
   if (isRepeat === false) {
     this.smtcTags.push({_id: smtcTagId});
+
+    if (!this.commentTo && this._media[0] === 'crowdtangle' && !this.url.match(/permalink/)) {
```
This is a lot of conditions; looks fine but seems ripe for a nice "why" comment in the code.

karpawich replied:
Comment added!

cooperka commented on public/angular/js/controllers/reports/show.js:
```diff
@@ -1,141 +1,205 @@
 angular.module('Aggie')
 
-.controller('ReportsShowController', [
+  .controller('ReportsShowController', [
```
Your IDE may have done this automatically, but everything else is unindented here so please undo this to keep things consistent.

cooperka commented on public/angular/js/controllers/reports/show.js:
```diff
@@ -1,141 +1,205 @@
+      $scope.saveReport = function(report) {
+        Report.save({ id: report._id }, report, function() {
+        }, function() {
+          flash.setAlertNow("Sorry, but that report couldn't be saved.");
```
Would be lovely to at least log an error to the console, lest it be swallowed by the void.

cooperka reviewed (commented):
Just some general code notes, I haven't tested this locally or looked at the HTML at all. Thank you for all this work!

Commit:
Only Show Comments Section if there are any comments.

karpawich committed:
tmp: add back fetch debugging

karpawich committed:
Merge branch 'max/comments-integration' of https://github.com/tid-lab/aggie into max/comments-integration

karpawich committed:
tmp: debug fetching #2

karpawich committed:
tmp: debug comments fetching #3

karpawich committed:
tmp: debug comment fetching #4

karpawich committed:
tmp: debug comment paging

karpawich committed:
fix: reset invalid JWT token

karpawich committed:
tmp: remove remaining console.log lines

harshilshah4251 committed:
hate speech integration

karpawich committed:
tmp: debug comment paging w/ hate speech

karpawich committed:
Revert "tmp: debug comment paging w/ hate speech"

This reverts commit 34673b78fa5c4584cbbb34662fa574beb36819c7.

Commit:
Updated Comment Time to Show Authored Time

It used to be the stored at time, which when the system stores the report value. This is incorrect.

Commit:
Merge branch 'develop' into max/comments-integration

karpawich committed:
fix: add back safeguard against multiple send() calls

karpawich committed:
fix: rename count to pageCount

karpawich committed:
chore: document verbose if statement

karpawich committed:
fix: replace callback hell w/ async/await

karpawich committed:
chore: add TODO notice for premature optimization

karpawich committed:
fix: callback plain reports if hate speech detection is disabled

Commit:
The Author Icon shouldn't Appear if Comment

Given we don't have authors for comments, remove the author if there isn't one.

Commit:
Merge branch 'develop' into max/comments-integration

Commit:
Added Read/Unread Flag/Unflag Incident functionality to comments

This also fixes the issue where the comments do not update upon adding incidents

Commit:
Fixed Report Controller Tag API

The report.save() function was only called in the error function callback so I just moved it outside.

Commit:
Removed Console.logs and Added Error Log

Commit:
Removed Edit Tags for Election

Commit:
Max forgot the case when tag is not Comment Tag

This case never called back

Commit:
Return Tag API endpoints in Report Controller to use Save in Callback
